### PR TITLE
Time - Yesenia

### PR DESCRIPTION
**Pull request by yesentorres**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
To instructors:  I would like to communicate that I acknowledge this submission is severely incomplete and that I should have had much more substantive work to show given all the time we had this week. I was dealing with a high level of stress this past week & weekend caused by external factors, specifically relating to the current health crisis in Washington, and other personal matters. This is what I have been able to accomplish and I would greatly appreciate an extension until the end of this week to complete the project. After having taken the time to work through my issues from last week, I should be able to focus on this and get it to pass Waves 1 & 2 at the very least. I also welcome a discussion of a learning plan if y'all think that is the right direction. Thank you. 
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | 
What was a design decision you made that changed over time over the project? | 
What was a concept you gained clarity on, or a learning that you'd like to share? | 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? |

### CODE REVIEW

File: lib/front_desk.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    def initialize(rooms: [], reservations: [])
+      @rooms = rooms
+      @reservations = reservations
+    end  
+    
+    def update_room_list(room)
+      @rooms << room 
+    end 
+
+    def create_rooms
```
Comment (beccaelenzil) [inline lib/front_desk.rb]: Great use of a helper method. I would consider using this method in the constructor so that when you initialize an instance of the FrontDesk class, you automatically get 20 rooms.

File: lib/front_desk.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    # find a room in @rooms array where the room's reservation's does not match the date range 
+    # returns: array of rooms where the supplied date range is NOT overlapping with any other of the room's reservation's date range
+    # modifies: nothing
+    def available_rooms(start_date, end_date)
+    end 
+
+    # create helper method that adds to @reservations array 
+    # returns: nothing 
+    # modifies: @reservations 
+    def update_reservations_list(reservation)
```
Comment (beccaelenzil) [inline lib/front_desk.rb:61]: This amount of work probably doesn't require it's own method (though it's not necessarily a bad thing to have one).

File: lib/reservation.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    attr_accessor :id, :room_num, :start_date, :end_date
+
+    def initialize(id:, room_num:, start_date:, end_date:)
+      @id = id
+      @room_num = room_num
+      @start_date = start_date
+      @end_date = end_date 
+    end 
+
+    def total_cost
```
Comment (beccaelenzil) [inline lib/reservation.rb:17]: Nice job working through this logic. Consider writing a separate method to calculate the duration of the stay and invoke this method when calculating the total cost.

File: lib/room.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+module Hotel 
+  class Room
+
+    attr_accessor :number, :cost, :reservation_ids
+
+    VALID_NUMS = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20] # hotel has 20 rooms
```
Comment (beccaelenzil) [inline lib/room.rb:6]: You can create this array with `(1..20).to_a`.

File: test/date_range_test.rb [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    it "raises an error if end_date is before start_date" do 
+      expect{Hotel::DateRange.new(start_date: Date.new(2020, 8, 9), end_date: Date.new(2020, 8, 4))}.must_raise ArgumentError
+    end
+
+    it "raises an error if start_date and end_date are same day" do 
+      expect{Hotel::DateRange.new(start_date: Date.new(2020, 8, 4), end_date: Date.new(2020, 8, 4))}.must_raise ArgumentError
+    end 
+  end
+
+  describe "overlap?" do
```
Comment (beccaelenzil) [inline test/date_range_test.rb:37]: Good work fleshing out these tests.

File: test/room_test.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      @test_room = Hotel::Room.new(number: 1)
+      expect(@test_room.cost).must_equal 200.00
+    end 
+
+    it "sets reservation_ids array to empty as default" do 
+      @test_room = Hotel::Room.new(number: 1)
+      expect(@test_room.reservation_ids.length).must_equal 0
+    end 
+
+    it "raises an error if valid room_id is not given" do
```
Comment (beccaelenzil) [inline test/room_test.rb:17]: Good work testing a failure case.

File: lib/date_range.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      if end_date < start_date 
+        raise ArgumentError.new("Invalid date range: End date, #{end_date}, is before  start date, #{start_date}")
+      elsif end_date == start_date
+        raise ArgumentError.new("Invalid date range: start_date and end_date is the same day")
+      end 
+      @start_date = start_date
+      @end_date = end_date
+    end 
+
+    def overlap?(other_date_range)
```
Comment (beccaelenzil) [inline lib/date_range.rb:16]: Nice work working through this complex logic.

**Comment by beccaelenzil** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️ Nice pseudo-code!
Practices git with at least 15 small commits and meaningful commit messages | 

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️ Yes. Still need functionality to create reservations.
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | Incomplete

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | Incomplete
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | Incomplete
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | Incomplete
gives back a list of available rooms for a given date range, and it has nominal test cases | Incomplete
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | Incomplete
creates a block of rooms | Incomplete
reserves a room from a block | Incomplete

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Yellow (Approaches Standards) | 9-13 total in all sections |
### Additional Feedback



You have a great start to this project. You've done a great job implementing the DateRange class and corresponding tests. The rest of your class design (architecture) has you set up for success. We will talk about a plan for completing (or not) during our one on one.

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Descriptive/Readable | ✅
| Logical/Organized | ✅

File: test/front_desk_test.rb [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+      expect(@test_desk.rooms.length).must_equal 20
+      expect(@test_desk.rooms[0]).must_be_kind_of Hotel::Room
+    end 
+
+    it "sets reservations list to empty arrays as default" do 
+      expect(@test_desk.reservations.length).must_equal 0 
+    end 
+  end 
+
+  xdescribe "all_rooms" do
```
Comment (beccaelenzil) [inline test/front_desk_test.rb:24]: Instead of having a method puts a value, you can have it return the value, and then puts it when the method is invoked. This is easier to test for.

File: test/front_desk_test.rb [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+    it "lists every reservation with a date range that includes the date being queried" do 
+      expect(@test_res_list.length).must_equal 3
+      expect(@test_res_list[0].id).must_equal 1
+      expect(@test_res_list[1].id).must_equal 3
+      expect(@test_res_list[2].id).must_equal 6
+    end 
+  end 
+
+  describe "reservations_by_room_and_range(room_num_query, range_query)" do
```
Comment (beccaelenzil) [inline test/front_desk_test.rb:59]: Nice, thorough tests.

File: test/front_desk_test.rb [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    
+    it "lists every reservation with a date range that is within the date range being queried" do 
+      expect(@test_res_list.length).must_equal 3
+      expect(@test_res_list[0].id).must_equal 3
+      expect(@test_res_list[1].id).must_equal 4
+      expect(@test_res_list[2].id).must_equal 5
+    end 
+  end
+
+  describe "available_rooms(range_query)" do
```
Comment (beccaelenzil) [inline test/front_desk_test.rb:94]: These tests look great. You also should test what is returned when there are no rooms available (perhaps an empty array or nil?)

File: test/front_desk_test.rb [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+      @test_desk = Hotel::FrontDesk.new()
+      start_date = Date.new(2020, 8, 1)
+      end_date = Date.new(2020, 8, 8)
+      i = 1 
+      # book all 20 rooms that conflict with date range provided 
+      20.times do
+        @test_desk.update_reservations_list(Hotel::Reservation.new(id: i, room_num: i, start_date: Date.new(2020, 8, 1), end_date: Date.new(2020, 8, 5))) 
+        i += 1
+      end  
+      expect{@test_desk.reserve_room(start_date, end_date)}.must_raise ArgumentError
```
Comment (beccaelenzil) [inline test/front_desk_test.rb:163]: Nice test of an edge case! Consider creating a custom exception for this situation.

**Review by beccaelenzil — COMMENTED** [review]
Great work completing this assignment! I found it really tricky to update the rubric, so I will put the additional comments here. I've added a few inline comments for you to review. The items still incomplete from the rubric are the comprehension questions (I think it would be helpful for your process to answer these) and edge case tests for the the `available_rooms` method. As you mentioned in our conversation, you FrontDesk methods made great use of your DateRange class. The code is well written and well tested. Keep up the hard work!
